File: app/src/main/java/com/barbershopsim/app/MainFragment.java
package com.barbershopsim.app;

import android.os.Bundle;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import com.barbershopsim.app.databinding.FragmentMainBinding;

import java.util.List;

public class MainFragment extends Fragment {
    private FragmentMainBinding binding;

    private MainViewModel mViewModel;

    private List<TextView> waitingCustomerViews, barberViews, barberCustomerViews;

    public static MainFragment newInstance() {
        return new MainFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mViewModel = new ViewModelProvider(this, new ViewModelProvider.NewInstanceFactory()).get(MainViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = FragmentMainBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        waitingCustomerViews =
                List.of(binding.waitingCustomerOne.getRoot(), binding.waitingCustomerTwo.getRoot(), binding.waitingCustomerThree.getRoot(), binding.waitingCustomerFour.getRoot());
        barberViews =
                List.of(binding.barberOne.getRoot(), binding.barberTwo.getRoot(), binding.barberThree.getRoot(), binding.barberFour.getRoot());
        barberCustomerViews =
                List.of(binding.barberCustomerOne.getRoot(), binding.barberCustomerTwo.getRoot(), binding.barberCustomerThree.getRoot(), binding.barberCustomerFour.getRoot());

        binding.shopName.setText(mViewModel.getShopName());
        binding.timescaleFactor.setText(mViewModel.getTimescaleFactor());

        mViewModel.isShopOpen.observe(getViewLifecycleOwner(), isShopOpen -> {
            if (isShopOpen == true) {
                binding.closedSign.setVisibility(View.INVISIBLE);
                binding.openSign.setVisibility(View.VISIBLE);
            } else {
                binding.closedSign.setVisibility(View.VISIBLE);
                binding.openSign.setVisibility(View.INVISIBLE);
            }
        });
        mViewModel.customerEnter.observe(getViewLifecycleOwner(), customer -> {
            if (customer == null) {
                binding.incomingCustomer.getRoot().setText("");
            } else {
                binding.incomingCustomer.getRoot().setText(String.format("%02d", customer));
            }
        });
        mViewModel.customerExit.observe(getViewLifecycleOwner(), customer -> {
            if (customer == null) {
                binding.outgoingCustomer.getRoot().setText("");
            } else {
                binding.outgoingCustomer.getRoot().setText(String.format("%02d", customer));
            }
        });
        mViewModel.waitingArea.observe(getViewLifecycleOwner(), this::updateWaitingArea);
        mViewModel.haircutArea.observe(getViewLifecycleOwner(), this::updateBarberChairs);
        mViewModel.clock.observe(getViewLifecycleOwner(), time -> binding.clock.setText(time));

        binding.restartButton.setOnClickListener(_ -> mViewModel.start());
        binding.playPauseButton.setOnClickListener(_ -> {
            // TODO Pause
            mViewModel.start();
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        waitingCustomerViews = barberViews = barberCustomerViews = null;
    }

    private void updateWaitingArea(List<Integer> customers) {
        for (var waitingCustomerView : waitingCustomerViews) {
            waitingCustomerView.setText("");
        }
        for (int i = 0; i < customers.size(); i++) {
            int customer = customers.get(i);
            waitingCustomerViews.get(i).setText(String.format("%02d", customer));
        }
    }

    private void updateBarberChairs(List<Pair<String, Integer>> chairs) {
        for (var barberView : barberViews) {
            barberView.setText("");
        }
        for (var barberCustomerView : barberCustomerViews) {
            barberCustomerView.setText("");
        }
        for (int i = 0; i < chairs.size(); i++) {
            Pair<String, Integer> chair = chairs.get(i);
            if (chair.first != null) {
                barberViews.get(i).setText(chair.first);
            }
            if (chair.second != -1) {
                barberCustomerViews.get(i).setText(String.format("%02d", chair.second));
            }
        }
    }
}

File: app/src/main/java/com/barbershopsim/app/MainFragment.kt
package com.barbershopsim.app

import android.os.Bundle
import android.util.Pair
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProvider.NewInstanceFactory
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.barbershopsim.app.databinding.FragmentMainBinding
import kotlinx.coroutines.launch

class MainFragment : Fragment() {
    companion object {
        fun newInstance(): MainFragment {
            return MainFragment()
        }
    }

    private var binding: FragmentMainBinding? = null

    private var viewModel: MainViewModel? = null

    private var waitingCustomerViews: List<TextView>? = null
    private var barberViews: List<TextView>? = null
    private var barberCustomerViews: List<TextView>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProvider(this, NewInstanceFactory())[MainViewModel::class.java]
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentMainBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val viewModel = viewModel!!
        val binding = binding!!

        waitingCustomerViews = listOf(
            binding.waitingCustomerOne.root,
            binding.waitingCustomerTwo.root,
            binding.waitingCustomerThree.root,
            binding.waitingCustomerFour.root
        )
        barberViews = listOf(
            binding.barberOne.root, binding.barberTwo.root, binding.barberThree.root, binding.barberFour.root
        )
        barberCustomerViews = listOf(
            binding.barberCustomerOne.root,
            binding.barberCustomerTwo.root,
            binding.barberCustomerThree.root,
            binding.barberCustomerFour.root
        )

        binding.shopName.text = viewModel.shopName
        binding.timescaleFactor.text = viewModel.timescaleFactor

        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModel.isShopOpen.collect { isShopOpen ->
                        if (isShopOpen) {
                            binding.closedSign.visibility = View.INVISIBLE
                            binding.openSign.visibility = View.VISIBLE
                        } else {
                            binding.closedSign.visibility = View.VISIBLE
                            binding.openSign.visibility = View.INVISIBLE
                        }
                    }
                }
                launch {
                    viewModel.customerEnter.collect { customer ->
                        customer?.let { binding.incomingCustomer.root.text = it.toString().padStart(2, '0') }
                            ?: run { binding.incomingCustomer.root.text = "" }
                    }
                }
                launch {
                    viewModel.customerExit.collect { customer ->
                        customer?.let { binding.outgoingCustomer.root.text = it.toString().padStart(2, '0') }
                            ?: run { binding.outgoingCustomer.root.text = "" }
                    }
                }
                launch {
                    viewModel.waitingArea.collect(::updateWaitingArea)
                }
                launch {
                    viewModel.haircutArea.collect(::updateBarberChairs)
                }
                launch {
                    viewModel.clock.collect { time ->
                        if (time.isNotEmpty()) {
                            binding.clock.text = time
                        }
                    }
                }
            }
        }

        binding.restartButton.setOnClickListener { viewModel.start() }
        binding.playPauseButton.setOnClickListener {
            // TODO Pause
            viewModel.start()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
        barberCustomerViews = null
        barberViews = null
        waitingCustomerViews = null
    }

    override fun onDestroy() {
        super.onDestroy()
        viewModel = null
    }

    private fun updateWaitingArea(customers: List<Int>) {
        for (waitingCustomerView in waitingCustomerViews!!) {
            waitingCustomerView.text = ""
        }
        for (i in customers.indices) {
            val customer = customers[i]
            waitingCustomerViews!![i].text = customer.toString().padStart(2, '0')
        }
    }

    private fun updateBarberChairs(chairs: List<Pair<String?, Int>>) {
        for (barberView in barberViews!!) {
            barberView.text = ""
        }
        for (barberCustomerView in barberCustomerViews!!) {
            barberCustomerView.text = ""
        }
        for (i in chairs.indices) {
            val chair = chairs[i]
            chair.first?.let { barberViews!![i].text = it }
            if (chair.second != -1) {
                barberCustomerViews!![i].text = chair.second.toString().padStart(2, '0')
            }
        }
    }
}